File: packages/lang-java/templates/Graph.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The campus as the player walks it.
 *
 * Two jobs: give the player something ergonomic to explore, and count. Every call to
 * {@code neighbors} is one node expanded, and that is the metric the game shows instead of
 * time (ADR 0011).
 */
final class Graph {
    /** One path leaving a building, with the distance to walk it. */
    record Edge(String to, double weight) {}

    private static int explored = 0;

    static void resetOps() {
        explored = 0;
    }

    static int opsCount() {
        return explored;
    }

    private final List<String> order = new ArrayList<>();
    private final Map<String, String> labels = new LinkedHashMap<>();
    private final Map<String, List<Edge>> adjacency = new LinkedHashMap<>();

    /** Every node id in the graph. */
    List<String> nodes() {
        return order;
    }

    /** The edges leaving a node. Each call counts as one node explored. */
    List<Edge> neighbors(String id) {
        explored++;
        return adjacency.getOrDefault(id, List.of());
    }

    /** The name a person reads, as opposed to the id an algorithm uses. */
    String label(String id) {
        return labels.getOrDefault(id, id);
    }

    static Graph from(Object raw) {
        Graph graph = new Graph();
        Map<String, Object> source = Json.asObject(raw);

        for (Object entry : Json.asList(source.get("nodes"), value -> value)) {
            Map<String, Object> node = Json.asObject(entry);
            String id = Json.asString(node.get("id"));
            graph.order.add(id);
            graph.labels.put(id, Json.asString(node.get("label")));
            graph.adjacency.put(id, new ArrayList<>());
        }

        // Edges are undirected and declared once, so both directions are filled in here.
        for (Object entry : Json.asList(source.get("edges"), value -> value)) {
            Map<String, Object> edge = Json.asObject(entry);
            String from = Json.asString(edge.get("from"));
            String to = Json.asString(edge.get("to"));
            double weight = Json.asDouble(edge.get("weight"));

            graph.adjacency.computeIfAbsent(from, key -> new ArrayList<>()).add(new Edge(to, weight));
            graph.adjacency.computeIfAbsent(to, key -> new ArrayList<>()).add(new Edge(from, weight));
        }

        return graph;
    }
}

File: packages/lang-elixir/templates/graph.ex
defmodule Edge do
  @moduledoc false

  @enforce_keys [:to, :weight]
  defstruct [:to, :weight]

  @type t :: %__MODULE__{to: String.t(), weight: float()}
end

defmodule Graph do
  @moduledoc """
  The campus as the player walks it.

  Two jobs: give the player something ergonomic to explore, and count. Every call to
  `neighbors/2` is one node expanded, and that is the metric the game shows instead of time
  (ADR 0011).

  The counter lives in the process dictionary. The BEAM has no mutable global, and this is
  the idiomatic way to keep one hidden inside a single process without it leaking into the
  player's own function signature.
  """

  @enforce_keys [:order, :labels, :adjacency]
  defstruct [:order, :labels, :adjacency]

  @type t :: %__MODULE__{
          order: [String.t()],
          labels: %{String.t() => String.t()},
          adjacency: %{String.t() => [Edge.t()]}
        }

  @ops_key :uq_graph_ops

  @spec reset_ops() :: :ok
  def reset_ops do
    Process.put(@ops_key, 0)
    :ok
  end

  @spec ops_count() :: non_neg_integer()
  def ops_count do
    Process.get(@ops_key, 0)
  end

  @spec new(map()) :: t()
  def new(raw) do
    nodes = Map.get(raw, "nodes", [])
    edges = Map.get(raw, "edges", [])

    order = Enum.map(nodes, & &1["id"])
    labels = Map.new(nodes, &{&1["id"], &1["label"]})

    # Edges are undirected and declared once, so both directions are filled in here.
    adjacency =
      Enum.reduce(edges, Map.new(order, &{&1, []}), fn edge, acc ->
        from = edge["from"]
        to = edge["to"]
        weight = edge["weight"]

        acc
        |> Map.update(from, [%Edge{to: to, weight: weight}], &(&1 ++ [%Edge{to: to, weight: weight}]))
        |> Map.update(to, [%Edge{to: from, weight: weight}], &(&1 ++ [%Edge{to: from, weight: weight}]))
      end)

    %__MODULE__{order: order, labels: labels, adjacency: adjacency}
  end

  @spec nodes(t()) :: [String.t()]
  def nodes(%__MODULE__{order: order}), do: order

  @spec neighbors(t(), String.t()) :: [Edge.t()]
  def neighbors(%__MODULE__{adjacency: adjacency}, id) do
    Process.put(@ops_key, ops_count() + 1)
    Map.get(adjacency, id, [])
  end

  @spec label(t(), String.t()) :: String.t()
  def label(%__MODULE__{labels: labels}, id), do: Map.get(labels, id, id)
end
